File: c++/HashTable/HashTable.cpp
#include <iostream>
#include <string>
#include <cstring>
#include "HashTable.h"
#include "HashNode.h"

using namespace std;

/* CONSTRUCTOR */
template <class T>
HashNode<T>::HashNode(T value, string key)
{
    this->value = value;
    this->key = key;
    this->next = NULL;
}

/* CONSTRUCTOR */
template <class T>
HashTable<T>::HashTable()
{
     // Starting capacity of 15
    this->capacity = 15;
    // Set size to 0
    this->size = 0;
    table = new HashNode<T> *[capacity];

    // Initialize the table
    for(int i = 0; i < capacity; i++)
    {
        table[i] = NULL;
    }
}

/* DESTRUCTOR */
template <class T>
HashTable<T>::~HashTable()
{
    // Go through each index of the table
    for(int i = 0; i < capacity; i++)
    {
        HashNode<T> *n = table[i];

        // Go through the list at each index and delete the node
        while(n != NULL)
        {
            HashNode<T> *prev = n;
            n = n->next;
            delete prev;
        }
    }

    // Delete the table
    delete [] table;
}

/**
 * Function to get the hash index of a key
 * 
 * @Param key           The key to find the hash index of
 * 
 * @Returns             The hash index of that key
 */
template<class T>
int HashTable<T>::hash(string key)
{
    char ch[key.length() + 1];
    strcpy(ch, key.c_str());

    int sum = 0;
    for(int i = 0; i < key.length(); i++)
    {
        sum += ch[i];
    }

    return sum % capacity;
}

/**
 * Function to insert a key value pair into the hash table
 * 
 * @Param key           The key of the pair to insert
 * @Param value         The value of the pair to insert
 */
template <class T>
void HashTable<T>::insert(string key, T value)
{
    // Get the hash index
    int hashInd = hash(key);

    // Get the list of the correct hash index
    HashNode<T> *temp = table[hashInd];
    HashNode<T> *prev = NULL;

    // Go through the list at that index
    while(temp != NULL)
    {
        prev = temp;
        temp = temp->next;
    }

    // If the last value is NULL, add new node in
    if(temp == NULL)
    {
        temp = new HashNode<T>(value, key);
        if(prev == NULL)
        {
            table[hashInd] = temp;
        }
        else
        {
            prev->next = temp;
        }
    }
    else
    {
        temp->value = value;
    }

    cout << "Value " << value << " at key " << key << " added." << endl;
    size++;
}

/**
 * Function to remove a pair with a certain key in the table (Removes the last value in the list at that key index)
 * 
 * @Param key           The key of the pair to remove
 */
template <class T>
void HashTable<T>::remove(string key)
{
    // Get the hash index of the key, and get the correct list from the table
    int hashInd = hash(key);
            
    // If there is a valid list at that index
    if(table[hashInd] != NULL)
    {
        // Get the list
        HashNode<T> *n = table[hashInd];
        HashNode<T> *prev = NULL;

        // Go through the list until found the key or at the end
        while(n->next != NULL && n->key.compare(key) != 0)
        {
            prev = n;
            n = n->next;
        }

        // If its the correct key
        if(n->key.compare(key) == 0)
        {
            // If the previous is NULL (the key found is the first one in the list), set the beginning of the list to the next
            if(prev == NULL)
            {
                HashNode<T> *nextNode = n->next;
                delete n;
                table[hashInd] = nextNode;
            }
            else
            {
                HashNode<T> *nextNode1 = n->next;
                delete n;
                prev->next = nextNode1;
            }
        }
    }

    cout << "Element at key \"" << key << "\" removed." << endl;
    size--;
}

/**
 * Function to get the value at a certain key in the map
 * 
 * @Param key           The key of the key value pair to look for
 * 
 */
template <class T>
T HashTable<T>::get(string key)
{
    // Get the hash index
    int hashInd = hash(key);

    // Get the list
    HashNode<T> *n = table[hashInd];
    bool found = false;

    // Go through the list until the key is found
    while(n != NULL && n->key.compare(key) != 0)
    {
        n = n->next;
    }

    // If the key was found, return it
    if(n->key.compare(key) == 0)
    {
        return n->value;
    }
    else
    {
        cout << "Entry with key " << key << " not found." << endl;
        return NULL;
    }
}

/**
 * Function to print the table
 */
template <class T>
void HashTable<T>::printTable()
{
    cout << endl << endl << "Printing table" << endl;;

    for(int i = 0; i < capacity; i++)
    {
        if(table[i] != NULL)
        {
            HashNode<T> *n = table[i];

            while(n != NULL)
            {
                cout << "Key: " << n->key << " Value: " << n->value << endl;
                n = n->next;
            }
        }
    }
    cout << endl << endl;
}


/***** TESTING *****/
int main()
{
    HashTable<string> *myTable = new HashTable<string>;
    myTable->insert("Ontario", "Canada");
    myTable->insert("New York", "USA");
    myTable->insert("Rome", "Italy");
    cout << "Size: " << myTable->sizeTable() << endl;
    cout << "Element at key \"Rome\": " << myTable->get("Rome") << endl;
    myTable->printTable();
    myTable->remove("New York");
    myTable->printTable();
    
    return 0;
}

File: python/HashTable/HashTable.py
class HashNode:
    # CONSTRUCTOR #
    def __init__(self, value = None, key = None):
        self.value = value
        self.key = key
        self.next = None
        return

class HashTable:
    # CONSTRUCTOR #
    def __init__(self):
        self.capacity = 15
        self.size = 0
        self.table = [None] * self.capacity

    # Function to  get the hash index from a key
    #
    # @Param key            The key to find the index for 
    #
    # @Returns              The hash index of that key
    def hash(self, key):
        total = 0
        for i in range(len(key)):
            total = total + ord(key[i])
        return total % self.capacity
    
    # Function to insert into the table
    #
    # @Param key            The key of the key/value pair
    # @Param value          The value of the key/value pair
    def insert(self, key, value):
        # Get hash index
        hashIndex = self.hash(key)

        # Get the list at the hashindex
        n = self.table[hashIndex]
        prev = None

        # Traverse through the list
        while n is not None:
            prev = n
            n = n.next

        # Add to the list
        if n is None:
            n = HashNode(value=value, key=key)

            if prev is None:
                self.table[hashIndex] = n
            else:
                prev.next = n
        else:
            n.value = value
        
        print("Value %s at key %s added." % (value, key))
        self.size = self.size + 1
    
    # Function to remove a key/value pair at a given key
    #
    # @Param key            The key of the pair to remove
    def remove(self, key):
        # Get the hash index
        hashIndex = self.hash(key)

        # If the list at that index is not empty
        if self.table[hashIndex] is not None:
            n = self.table[hashIndex]
            prev = None

            # Traverse through the list until found the key
            while(n.next is not None and n.key != key):
                prev = n
                n = n.next
            
            # If found the key in the list, remove the pair
            if n.key == key:
                if prev is None:
                    self.table[hashIndex] = n.next
                else:
                    prev.next = n.next
                
            print("Element at key \"%s\" removed." % (key))
            self.size = self.size - 1
        
    # Function to get the value of a given key
    #
    # @Param key            The key to find the value of
    #
    # @Returns              The value of the given key
    def get(self, key):
        # Get hash index
        hashIndex = self.hash(key)

        n = self.table[hashIndex]
        found = False

        # Traverse through the list until the end or the key is found
        while n is not None and n.key != key:
            n = n.next

        # If the key was found, return
        if n.key == key:
            return n.value
        else:
            print("Entry with key \"%s\" not found." % (key))
            return

    # Function to print the table out
    def printTable(self):
        print()
        print()
        print("Printing table")

        for i in range(self.capacity):
            if self.table[i] is not None:
                n = self.table[i]

                while n is not None:
                    print("Key: %s Value: %s." % (n.key, n.value))
                    n = n.next
                
        print()
        print()
    
    # Function to return the size of the table
    #
    # @Returns          The size of the table
    def sizeTable(self):
        return self.size

    # Function to check if the table is empty
    #
    # @Returns          True or false, Empty or not
    def isEmpty(self):
        if self.size == 0:
            return True

        return False
    
###### TESTING ######
myTable = HashTable()
myTable.insert("Ontario", "Canada")
myTable.insert("New York", "USA")
myTable.insert("Rome", "Italy")
print("Size: %s" % (myTable.sizeTable()))
print("Element at key \"Rome\": %s" % (myTable.get("Rome")))
myTable.printTable()
myTable.remove("New York")
myTable.printTable()
            
